### PR TITLE
Raising proposal based on PR discussion for: Enhance KafkaBridge resurce with consumer inactivity timeout and HTTP consumer/producer parts enablement

### PR DESCRIPTION
**Pull request by steffen-karlsson**

### CODE REVIEW

File: 068-kafkabrige-consumer-producer.md
Comment (scholzj) [inline 068-kafkabrige-consumer-producer.md:2]: You can delete this I guess.

File: 068-kafkabrige-consumer-producer.md
Comment (scholzj) [inline 068-kafkabrige-consumer-producer.md:47]: The CRD structure is not really much useful for the review as it is hard to read. It would be much more useful to include examples of a KafkaBridge resource with the new API fields to show how it will look like.

File: 068-kafkabrige-consumer-producer.md
Comment (scholzj) [inline 068-kafkabrige-consumer-producer.md:20]: Ideally, the proposal should be readable as a single document -> can you maybe summarize the things from the comments instead of linking to them?

**Review by scholzj — COMMENTED** [review]
Thanks for the proposal. I left few initial comments.

File: 068-kafkabrige-consumer-producer.md
Comment (steffen-karlsson) [reply 068-kafkabrige-consumer-producer.md:47]: Will work on that and update

**Comment by steffen-karlsson** [comment]
> Thanks for the proposal. I left few initial comments.

Comments resolved @scholzj.

File: 068-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 068-kafkabrige-consumer-producer.md:6]:
```suggestion
* `http.producer.enabled` - To enable/disable the HTTP producer part (enabled by default).
```

File: 068-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 068-kafkabrige-consumer-producer.md:69]: I prefer Version 2 and it also lets me think that we made a mistake in the past to put enablement of producer and consumer under the `http.` section. In the end, producer and consumer are an intrinsic part of the bridge or anyway when enabling/disabling one of them you are actually modifying the overall behaviour for both. I mean, you are enabling/disabling both HTTP and Kafka part of each. There is no distinction.

Comment (steffen-karlsson) [reply 068-kafkabrige-consumer-producer.md:69]: Maybe it should be called `httpEnabled` instead of just `enabled` then?

If you'd like, I can fix this in the operator API and then we can look at moving it inside the Bridge after?

Comment (ppatierno) [reply 068-kafkabrige-consumer-producer.md:69]: No I think that `enabled` is the right word to use because I don't see it totally related to HTTP protocol.
Also there are no changes to make on the PR. It has to be parked. We need to reach an agreement on the API through the proposal with all interested maintainers.
Even because just because I like Version 2, it doesn't mean we'll go with Version 2 :-) ... we need to wait what the others think.

File: 068-kafkabrige-consumer-producer.md
Comment (scholzj) [inline 068-kafkabrige-consumer-producer.md:75]: So, I guess the affected projects are the HTTP Bridge (changes to the start-up script) and the operators project where you extend the API and the Cluster Operator.

File: 068-kafkabrige-consumer-producer.md
Comment (scholzj) [inline 068-kafkabrige-consumer-producer.md:79]: You can add that there is no impact on backward compatibility or something like that.

**Review by scholzj — COMMENTED** [review]
Maybe I should try to explain the proposal process a bit ... the proposal is used to discuss, design, and decide how some more complicated issues will be addressed. Thsi includes typically things that:
* Change the API (the CRDs)
* Might have some techncial consequences, cause future limitations etc.

In this case, the change of the API / CRDs would be the main reason for the proposal. While it is sometimes useful to have a prototype PR written in is not necessarily required. I personally would not link it in the proposal but for example in the comments / PR description as an example of how it might be implemented.

If there are multiple variants -> you would normally choose one and propose it and list the others in the rejected alternatives for others to consider. This could be based on your liking but also on the follow-up discussion. So in this case we should pick one version and use it and move the other to the rejected alternatives. It seems the version two is prefereed by Paolo and probbaly slighly by me as well. So maybe we can pick that one?

Having the one variant selected is important, because at the end, the proposal is voted upon and needs to get at least 3 maintainer votes (approvals) (see more info [here](https://github.com/strimzi/governance/blob/main/GOVERNANCE.md#proposals)). So it needs to be clear what the plan is.

So we should pick the right variant, get the votes which I think should be pretty straigth forward here and then we can proceed with the PRs.

**Comment by steffen-karlsson** [comment]
@scholzj, thanks for explaining :)
Though it was merged and then discussed after, but that makes sense - my bad.

File: 068-kafkabrige-consumer-producer.md
Comment (scholzj) [inline 068-kafkabrige-consumer-producer.md:34]: I guess this (and the producer one) defaults to true for backward compatibility? It would be worth mentioning it in the text either here or in the compatibility section.

Comment (steffen-karlsson) [reply 068-kafkabrige-consumer-producer.md:34]: Sure, that makes sense :)

**Comment by scholzj** [comment]
@k-wall I would try to avoid doing the deep nesting and having `consumer` sections everywhere. So I would not put it under `.soec.http.consumer` and should it be related only to consumer, I would go for `.spec.consumer.timeoutSeconds`. WDYT @ppatierno?

**Comment by steffen-karlsson** [comment]
@scholzj if `timeoutSeconds` should be available for both `consumer` and `producer`, I think thats a better solution, more in line with version 2 presented.

**Comment by scholzj** [comment]
> @scholzj if `timeoutSeconds` should be available for both `consumer` and `producer`, I think thats a better solution, more in line with version 2 presented.

If it applies to both, it makes more sense under `spec.http` as you have it I guess. That also coresponds to `http.timeoutSeconds` as it exists in the bridge itself. But I would defer to @ppatierno as to what he thinks.

**Comment by steffen-karlsson** [comment]
> > @scholzj if `timeoutSeconds` should be available for both `consumer` and `producer`, I think thats a better solution, more in line with version 2 presented.
> 
> If it applies to both, it makes more sense under `spec.http` as you have it I guess. That also coresponds to `http.timeoutSeconds` as it exists in the bridge itself. But I would defer to @ppatierno as to what he thinks.

Sure thing. Sorry for the potential misunderstanding, what I meant was if its relevant for those `timeoutSeconds` to differ from `consumer` to `producer`.

**Comment by k-wall** [comment]
https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/java/io/strimzi/kafka/bridge/http/HttpConfig.java#L37 shows this property applies only to the consumer.  I there should be a change considered against the Bridge itself that would rename the property `http.consumer.timeoutSeconds` (retaining but deprecating the old name). The API being proposed by this change should be `.spec.consumer.timeoutSeconds`.

**Comment by ppatierno** [comment]
The choice is not so obvious :-( and I think that the initial choice of `http.timeoutSeconds` was bad (sorry for that), because it didn't make it clear that the configuration is for the consumer only (as pointed by @k-wall).
The thing is that the bridge has two parts: http and kafka. But they are not that clear in the `KafkaBridge` CRD (i.e. `consumer.config` is about Kafka related parameter so everything like `kafka.*`, but it's not `consumer.kafka.config` which could have been better).
Another idea ... I know it's going to use a different order from the application.properties file but what about using `consumer.http.timeoutSeconds`, `consumer.http.enabled`, `producer.http.enabled` ? I know they will be the opposite of what we have in the application properties starting with `http.*` but we already have the approach of `consumer` and `producer` sections in the `KafkBridge` custom resource.
At the same time it means an additional inner level, which we could avoid.
@scholzj @k-wall @steffen-karlsson wdyt?

**Comment by scholzj** [comment]
@ppatierno I would try to avoid the deep nesting and the confusing use of `http` in multiple places. So maybe you could do something like:

```
spec:
  http:
    consumerEnabled: true
    producerEnabled: true
    consumerTimeoutSeconds: 60
```

if you think these things belong to HTTP?

**Comment by ppatierno** [comment]
@scholzj also your suggestion is fine but I am really struggling to take a strong position. I am really regretting the choices made in the past for the current names in the application.properties file.
Regarding the `enabled` field for example, while it's enabling/disabling the HTTP part of consumer/producer, in reality it's enabling/disabling the consumer/producer as whole (disabling HTTP, no Kafka client is created) so the best should have been just `consumer.enabled` and `producer.enabled` (without the `http` prefix). So the timeout seconds should have been just consumer (if we assume that any other Kafka related timeout should go in the config part).
At this point I would be for having these params directly under consumer (enabled and timeoutSeconds) and producer (enabled) sections.

**Comment by scholzj** [comment]
APIs are hard :-(. When it comes to bad decisions in the `KafkaBridge` custom resource, I guess the `v1` api will be an opportunity. To change the `application.properties` is I guess less complicated technically, but would break compatibility. But in any case I don't think the `KafkaBridge` schema absolutely has to follow the `application.properties` if it doesn't make sense.

**Comment by k-wall** [comment]
Isn't calling it a http timeout is actually misleading?   It could be confused with the timeouts applied to the HTTP protocol itself.   The timeout we are talking about is intrinsic to the consumer bridge.  So that train of thought leads me to:

```
consumer:
    enabled: true
    timeoutSeconds: 60
    config:
      auto.offset.reset: earliest
      enable.auto.commit: true
```

**Comment by steffen-karlsson** [comment]
@ppatierno Sure, let me do that :)

File: 071-kafkabrige-consumer-producer.md
Comment (k-wall) [inline 071-kafkabrige-consumer-producer.md:18]: This sentence is now inconsistent with the updated API proposal.

File: 071-kafkabrige-consumer-producer.md
Comment (k-wall) [inline 071-kafkabrige-consumer-producer.md:38]: I don't think producer actually has the concept of timeout.  It is only the consumer that does.

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [reply 071-kafkabrige-consumer-producer.md:18]: +1 for @k-wall 
Also I would avoid the code formatting on "http enablement", "consumer" and "producer" but using it for actual code and fields.

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [reply 071-kafkabrige-consumer-producer.md:38]: +1 It has not, you are right.

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:84]: again `timeoutSeconds` is valid for consumer only.

**Review by ppatierno — CHANGES_REQUESTED** [review]

**Comment by steffen-karlsson** [comment]
Not sure exactly what you mean by "Also I would avoid the code formatting on "http enablement", "consumer" and "producer" but using it for actual code and fields." @ppatierno

Everything else is updated @ppatierno and @k-wall.

**Comment by ppatierno** [comment]
> Not sure exactly what you mean by "Also I would avoid the code formatting on "http enablement", "consumer" and "producer" but using it for actual code and fields." @ppatierno

The usage of the single apex which provides the formatting like this `formatting with single apex` is typically for code and yaml fields while you are just mentioning "http enablement", "consumer" and "producer". So I would avoid having something like `http enablement`, `consumer` and `producer` unless you refer to fields in the spec or code.

File: 071-kafkabrige-consumer-producer.md
Comment (k-wall) [inline 071-kafkabrige-consumer-producer.md:50]: for thoroughness, we should talk here about compatibility aspects of `timeoutSeconds`.  If field is omitted from CR, I would expect the effect on the bridge would be as if `http.consumer.timeoutSeconds` had the value [`-1`](https://github.com/strimzi/strimzi-kafka-bridge/blob/25b4106bd5cf3b9b4c8f128b6c284159bb736149/src/main/java/io/strimzi/kafka/bridge/http/HttpConfig.java#L50).

Comment (ppatierno) [reply 071-kafkabrige-consumer-producer.md:50]: Yes when you omit `http.consumer.timeoutSeconds` from the application.properties, it takes `-1` as default value (no timeout). Missing `timeoutSeconds` in the CR would mean not setting it in the application.properties so providing the same effect. Of course, I agree it should be clarified in this section.

**Comment by ppatierno** [comment]
@steffen-karlsson maybe I was not clear but now your removed single apex from everywhere even where it makes sense :-D ... for example `http.consumer.timeoutSeconds` makes sense having the single apex and formatted as code, but you removed from it having just http.consumer.timeoutSeconds.
As I said, YAML fields should have such format.

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:4]:
```suggestion
* `http.consumer.timeoutSeconds`   - For deleting inactive consumers after a timeout (disabled by default).
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:5]:
```suggestion
* `http.consumer.enabled`          - To enable/disable the HTTP consumer part (enabled by default).
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:6]:
```suggestion
* `http.producer.enabled`          - To enable/disable the HTTP producer part (enabled by default).
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:18]:
```suggestion
This proposal suggests based on discussions here: https://github.com/strimzi/strimzi-kafka-operator/pull/9820 to add the http enablement for the consumer and producer in their respective sessions, i.e. `spec.consumer.enabled` and `spec.producer.enabled`, as well as `spec.consumer.timeoutSeconds` property.
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:49]:
```suggestion
Not specifying `spec.consumer.enabled` or `spec.producer.enabled` implies not configuring the corresponding bridge parameters `http.consumer.enabled` and `http.producer.enabled` which default to true, thus there are no impacts on backward compatibility.
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:51]: It seems you are referring to the property you would have in the `KafkaBridge` which is omitted but it's not `http.consumer.timeoutSeconds` but it's `spec.consumer.timeoutSeconds`.

```suggestion
Not specifying `spec.consumer.timeoutSeconds` implies not configuring the corresponding bridge parameter `http.consumer.timeoutSeconds` which defaults to -1 to have the same effect (no timeout), thus there are no impacts on backward compatibility.
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:56]:
```suggestion
Initially in the implementation suggestion: https://github.com/strimzi/strimzi-kafka-operator/pull/9820, the enabled properties for producer and consumer were added to as part of the http properties, i.e. `spec.http.consumer.enabled` and `spec.http.producer.enabled`.
```

File: 071-kafkabrige-consumer-producer.md
Comment (ppatierno) [inline 071-kafkabrige-consumer-producer.md:85]: I think we can remove this sentece. It's just part of the proposed solution.

**Review by ppatierno — CHANGES_REQUESTED** [review]

**Review by scholzj — APPROVED** [review]
I think Paolo's comments make sense and should be applied. But otherwise it looks good to me.

**Comment by ppatierno** [comment]
@steffen-karlsson just adding one more thing. The `071` number was just taken by another merged proposal.
You should rename the file using the next available number and also update the README file with this proposal (number and title).

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks for the proposal!

**Review by im-konge — APPROVED** [review]

File: 072-kafkabrige-consumer-producer.md
Comment (see-quick) [inline 072-kafkabrige-consumer-producer.md:45]:
```suggestion
Strimzi Kafka Bridge - A PR has been raised for the HTTP Bridge to use the new ENV parameters introduced by the operator:https://github.com/strimzi/strimzi-kafka-bridge/pull/882
```

**Review by see-quick — APPROVED** [review]
LGTM

**Review by k-wall — APPROVED** [review]

**Comment by ppatierno** [comment]
@steffen-karlsson as for the other PR, I see a lot of commits which are not related to the proposal. What's happened? Can you try to clean it please? Closing and opening is not good because we'll lost all the approvals we already have so you should fix on this PR imho.

**Comment by steffen-karlsson** [comment]
> @steffen-karlsson as for the other PR, I see a lot of commits which are not related to the proposal. What's happened? Can you try to clean it please? Closing and opening is not good because we'll lost all the approvals we already have so you should fix on this PR imho.

Git happend, tried to update from main due to conflict and then my fork was out of sync.
Let me squash it all into one

**Comment by ppatierno** [comment]
This proposal has enough votes to be approved and merged (4 binding, 1 non-binding).
Because it's has been opened for long time, I will wait the end of tomorrow (April 16th) and then approve/merged if no further feedback are provided.

**Comment by ppatierno** [comment]
This proposal is approved with 4 binding and 1 non-binding votes. Going to merge.
@steffen-karlsson you can even back to work on the implementation PR changing it accordingly. Thanks!

**Pull request merged**
